src/string.rs

```diff
@@ -0,0 +1,161 @@
+// really can't be bothered right now to write docs before I even wrote unit tests
+#![allow(missing_docs)]
+
+use binrw::binrw;
+// TODO we might want to consider automating this offsetting back
+// forth away by using something the the modular-bitfield crate
+const MAX_SHORTSTR_SIZE: usize = ((u8::MAX >> 1) - 1) as usize;
+
+fn only_contains_ascii(string: &str) -> bool {
+    string.chars().all(|c| c.is_ascii())
+}
+/// A String encapsulating how DeviceSQL VARCHAR's are structured
+///
+/// This implementation forces them to be immutable for now
+#[derive(Debug, PartialEq)]
+#[binrw]
+enum DeviceSQLString {
+    /// Short-String optimization case
+    ShortASCII(DeviceSQLShortASCIIString),
+    /// Regular long form strings, containing possibly different encodings
+    Long(DeviceSQLLongString),
+}
+
+impl DeviceSQLString {
+    /// Initializes a DeviceSQLString from a plain Rust String
+    fn new(string: String) -> Self {
+        let len = string.len();
+        let only_ascii = only_contains_ascii(&string);
+        if only_ascii && len <= MAX_SHORTSTR_SIZE {
+            let header = (((len + 1) as u8) // compensate for header byte
+            << 1) // pack length
+            | 1; // set LSB flag to indicate ShortASCII
+            Self::ShortASCII(DeviceSQLShortASCIIString {
+                header,
+                content: string,
+            })
+        } else if only_ascii && len <= (i16::MAX as usize) {
+            Self::Long(DeviceSQLLongString {
+                flags: 0x40,
+                length: (len + 4) as u16,
+                content: LongBody::Ascii(string),
+            })
+        } else if len <= (i16::MAX as usize) {
+            Self::Long(DeviceSQLLongString {
+                flags: 0x90,
+                length: (len + 4) as u16, // TODO this length totally wrong
+                content: LongBody::Ucs2le(string),
+            })
+        } else {
+            todo!()
+        }
+    }
+}
+
+#[derive(Debug, PartialEq)]
+#[binrw]
+struct DeviceSQLShortASCIIString {
+    #[br(assert(header & 0b1 == 1))]
+    header: u8,
+    #[br(count = (header >> 1) - 1)]
+    #[br(try_map = String::from_utf8)]
+    #[bw(map = |x: &String| x.as_bytes())]
+    content: String,
```

Comment:
`NullString` uses a `Vec<u8>` internally and then provides `into_string`/`into_string_lossless` methods. Any particular reason why you chose a different approach? https://docs.rs/binrw/latest/src/binrw/strings.rs.html#57-60

Reply:
Yeah, I only found out about that after the fact, I'll have to look into the API ergonomics, I might consider adopting it.